### DISCUSSION TITLE
Call or use mutation inside middleware or baseQuery

### DISCUSSION THREAD

**Discussion by shurshilov** [reduxjs/redux-toolkit#4153] [Q&A] [answered]
I wanted to implement reauthentication or reconnect when the token (session) expired.
 By default, this is 7 days, and in order not to re-enter credits after 7 days, just authenticate again (in hidden mode). 
The server is able to respond that the session has expired. 
I had 2 possible solutions to the problem. 
1 intercept an error inside middleware and if it is a session expiration error, then re-login will be called. 
2 do the same inside the baseQuery. In the documentation, I found a solution through baseQuery. 
BUT it doesn't work in the latest version. Since the code is of the type
```
const promise = api.dispatch(api.endpoints.login.initiate());
const { data, isLoading, isSuccess /*...*/ } = await promise;
```
it doesn't work because there are no endpoints in the api object.
The same is true in the middleware object, the api does not contain endpoints
as result i duplicate code from login ,utation like this
![image](https://github.com/reduxjs/redux-toolkit/assets/11828278/9eea7e69-4d70-4970-826a-fe110cf71956)
and ,y mutation look like this
![image](https://github.com/reduxjs/redux-toolkit/assets/11828278/93f2e908-52a9-4446-880b-09014de9177a)
I am not sure this is BUG or what can you help me...

**Comment by phryneas** [comment]
Are you injecting the auth mutation?

**Reply by shurshilov** [reply]
Hi, thank you for quick reply. Yes its true
![image](https://github.com/reduxjs/redux-toolkit/assets/11828278/666675be-875b-4842-86ef-f0fb4c039cb9)
I also do a simple import, I think it should work. Although I haven't tested it yet. But it would probably be good to add a couple of lines about this in the documentation. Because I studied about 50 projects on github who used middleware and realized that no developer caused mutations in middleware. Free analysis;) but I haven't analyzed baseQuery usage
![image](https://github.com/reduxjs/redux-toolkit/assets/11828278/77806e3b-21de-40fc-9d7f-ac0f531ce190)

**Reply by EskiMojo14** [reply, marked as answer]
https://redux-toolkit.js.org/rtk-query/usage/code-splitting

> Calling injectEndpoints will inject the endpoints into the original API, but also give you that same API with correct types for these endpoints back. (Unfortunately, it cannot modify the types for the original definition.)

**Reply by shurshilov** [reply]
But this is a bit not obvious, because I use a hook that is set in authApi and therefore the error interception should be in the context of authApi, which has endpoints set?
![image](https://github.com/reduxjs/redux-toolkit/assets/11828278/32264718-318a-42f3-9c25-3f87bc20cb59)
![image](https://github.com/reduxjs/redux-toolkit/assets/11828278/9012d55b-5f41-4d62-811c-47b1c1b03c7e)
![image](https://github.com/reduxjs/redux-toolkit/assets/11828278/fa5d1429-1f8f-4bf0-8616-78923b5bf121)

**Reply by phryneas** [reply]
But TypeScript doesn't have any way of knowing that X other things happened before that line of code is reached. TypeScript does a static analysis and it's just not there yet.

**Reply by shurshilov** [reply]
The problem is not that typescript "does not see" endpoints. The problem is that they really aren't there

**Reply by EskiMojo14** [reply]
right, that's the other thing - if this code executes before the injectEndpoints call, then the endpoints really aren't there.

therefore you should import the authApi version *so* the injectEndpoints call happens first
